File: ios/LocationSyncer.swift
import Foundation

/// Mirrors `SyncResult` in the TypeScript layer.
struct SyncResultModel {
  let success: Bool
  let count: Int
  let status: Int?
  let error: String?

  func toDictionary() -> [String: Any?] {
    return ["success": success, "count": count, "status": status, "error": error]
  }

  /// Nitro struct for the JS bridge. Numeric fields are `Double` per the
  /// generated `NativeSyncResult` initializer.
  func toNitro() -> NativeSyncResult {
    return NativeSyncResult(
      success: success,
      count: Double(count),
      status: status.map { Double($0) },
      error: error
    )
  }
}

/// Drains the ``LocationBufferStore`` to a remote endpoint in FIFO batches over
/// `URLSession`. Rows are deleted only after a 2xx, giving at-least-once
/// delivery. A lock enforces single-flight so an auto-sync tick and a manual
/// `flush()` never double-send.
final class LocationSyncer {
  static let shared = LocationSyncer()

  private let queue = DispatchQueue(label: "rn.pbl.sync")
  private let lock = NSLock()
  private var flushing = false
  private let maxBatchesPerFlush = 20

  func flush(_ config: LocationConfig, completion: @escaping (SyncResultModel) -> Void) {
    lock.lock()
    if flushing {
      lock.unlock()
      completion(SyncResultModel(success: false, count: 0, status: nil, error: "busy"))
      return
    }
    flushing = true
    lock.unlock()

    queue.async {
      let result = self.flushSync(config)
      self.lock.lock()
      self.flushing = false
      self.lock.unlock()
      completion(result)
    }
  }

  private func flushSync(_ config: LocationConfig) -> SyncResultModel {
    guard let urlString = config.syncUrl, let url = URL(string: urlString) else {
      return SyncResultModel(success: false, count: 0, status: nil, error: "no-sync-url")
    }

    let store = LocationBufferStore.shared
    var totalSent = 0
    var batches = 0

    while batches < maxBatchesPerFlush {
      let batch = store.takeBatch(config.batchSize)
      if batch.isEmpty { break }

      let payload = batch.map { $0.1.toJSONObject() }
      let res = post(url: url, config: config, payload: payload, count: batch.count)
      if !res.success {
        // Leave this batch (and the rest) queued for the next attempt.
        return totalSent > 0
          ? SyncResultModel(success: true, count: totalSent, status: res.status, error: nil)
          : res
      }

      store.deleteIds(batch.map { $0.0 })
      totalSent += batch.count
      batches += 1
    }

    return SyncResultModel(
      success: true,
      count: totalSent,
      status: totalSent > 0 ? 200 : nil,
      error: nil
    )
  }

  private func post(url: URL, config: LocationConfig, payload: [[String: Any]], count: Int) -> SyncResultModel {
    guard let body = try? JSONSerialization.data(withJSONObject: payload, options: []) else {
      return SyncResultModel(success: false, count: 0, status: nil, error: "encode-failed")
    }

    var request = URLRequest(url: url)
    request.httpMethod = config.httpMethod.uppercased() == "PUT" ? "PUT" : "POST"
    request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
    request.setValue("application/json", forHTTPHeaderField: "Accept")
    for (key, value) in config.headers {
      request.setValue(value, forHTTPHeaderField: key)
    }
    request.httpBody = body
    request.timeoutInterval = 20

    let semaphore = DispatchSemaphore(value: 0)
    var result = SyncResultModel(success: false, count: 0, status: nil, error: "unknown")

    let task = URLSession.shared.dataTask(with: request) { _, response, error in
      defer { semaphore.signal() }
      if let error = error {
        result = SyncResultModel(success: false, count: 0, status: nil, error: error.localizedDescription)
        return
      }
      guard let http = response as? HTTPURLResponse else {
        result = SyncResultModel(success: false, count: 0, status: nil, error: "no-response")
        return
      }
      if (200..<300).contains(http.statusCode) {
        result = SyncResultModel(success: true, count: count, status: http.statusCode, error: nil)
      } else {
        result = SyncResultModel(success: false, count: 0, status: http.statusCode, error: "http-\(http.statusCode)")
      }
    }
    task.resume()
    // Bounded wait so a hung connection can't wedge the sync queue forever.
    _ = semaphore.wait(timeout: .now() + 30)
    return result
  }
}

File: android/src/main/java/com/margelo/nitro/persistentbackgroundlocation/LocationSyncer.kt
package com.margelo.nitro.persistentbackgroundlocation

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import org.json.JSONArray
import java.io.BufferedOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.zip.GZIPOutputStream

/**
 * Drains the [LocationBufferStore] to a remote endpoint in FIFO batches.
 *
 * Runs entirely natively (no JS, no Play Services) so it keeps working while the
 * app is killed. Rows are deleted only after the server acknowledges them with a
 * 2xx, giving at-least-once delivery semantics. A [Mutex] enforces single-flight
 * so an auto-sync tick and a manual `flush()` never double-send.
 */
internal class LocationSyncer(
  private val store: LocationBufferStore,
  private val debug: Boolean
) {
  private val mutex = Mutex()

  /**
   * Flush as much of the buffer as possible. Stops at the first failed batch (so
   * a flaky network leaves the rest queued) and at [MAX_BATCHES_PER_FLUSH] to
   * bound a single call's duration.
   */
  suspend fun flush(config: LocationConfig): SyncResult = mutex.withLock {
    val url = config.syncUrl
    if (url.isNullOrEmpty()) {
      return@withLock SyncResult(success = false, count = 0, status = null, error = "no-sync-url")
    }

    withContext(Dispatchers.IO) {
      var totalSent = 0
      var batches = 0
      var lastStatus: Int? = null
      while (batches < MAX_BATCHES_PER_FLUSH) {
        val batch = store.takeBatch(config.batchSize)
        if (batch.isEmpty()) break

        val payload = JSONArray().apply {
          batch.forEach { (_, fix) -> put(fix.toJson()) }
        }.toString()

        val result = postBatch(url, config, payload, batch.size)
        if (!result.success) {
          // Leave this batch (and the rest) in the buffer for the next attempt.
          // On a partial flush report `status = null` rather than the failed
          // batch's code, so `success = true` never travels with a 4xx/5xx.
          return@withContext if (totalSent > 0) {
            SyncResult(success = true, count = totalSent, status = lastStatus, error = null)
          } else {
            result
          }
        }

        lastStatus = result.status
        store.deleteIds(batch.map { it.first })
        totalSent += batch.size
        batches++
      }
      SyncResult(success = true, count = totalSent, status = if (totalSent > 0) lastStatus else null, error = null)
    }
  }

  private fun postBatch(
    urlString: String,
    config: LocationConfig,
    payload: String,
    count: Int
  ): SyncResult {
    var connection: HttpURLConnection? = null
    return try {
      connection = (URL(urlString).openConnection() as HttpURLConnection).apply {
        requestMethod = if (config.httpMethod.equals("PUT", ignoreCase = true)) "PUT" else "POST"
        connectTimeout = CONNECT_TIMEOUT_MS
        readTimeout = READ_TIMEOUT_MS
        doOutput = true
        setRequestProperty("Content-Type", "application/json; charset=utf-8")
        setRequestProperty("Accept", "application/json")
        setRequestProperty("Content-Encoding", "gzip")
        config.headers.forEach { (k, v) -> setRequestProperty(k, v) }
      }

      val bytes = payload.toByteArray(Charsets.UTF_8)
      BufferedOutputStream(connection.outputStream).use { raw ->
        GZIPOutputStream(raw).use { gzip -> gzip.write(bytes) }
      }

      val status = connection.responseCode
      if (status in 200..299) {
        if (debug) Log.d(Constants.TAG, "Synced $count fix(es) → HTTP $status")
        SyncResult(success = true, count = count, status = status, error = null)
      } else {
        val message = runCatching {
          connection.errorStream?.bufferedReader()?.use { it.readText() }
        }.getOrNull()?.take(256)
        if (debug) Log.w(Constants.TAG, "Sync failed → HTTP $status: $message")
        SyncResult(success = false, count = 0, status = status, error = message ?: "http-$status")
      }
    } catch (t: Throwable) {
      if (debug) Log.w(Constants.TAG, "Sync request threw", t)
      SyncResult(success = false, count = 0, status = null, error = t.message ?: t.javaClass.simpleName)
    } finally {
      connection?.disconnect()
    }
  }

  companion object {
    private const val CONNECT_TIMEOUT_MS = 15_000
    private const val READ_TIMEOUT_MS = 20_000
    private const val MAX_BATCHES_PER_FLUSH = 20
  }
}
